### PR TITLE
Dev/xygu/20200825/chips

### PR DESCRIPTION
**Pull request by Xiaoy312**
﻿GitHub Issue: #n/a

## PR Type

What kind of change does this PR introduce?

- Feature

## Description

Added Implementation for Chip & ChipGroup controls.

## PR Checklist 
Please check if your PR fulfills the following requirements:

- Commits must be following the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [x] Tested UWP
- [ ] Tested iOS
- [x] Tested Android
- [x] Tested WASM
- [ ] Tested MacOS
- [x] Contains **No** breaking changes
  > If the pull request contains breaking changes, commit message must contain a detailed description of the action to take for the consumer of this library. As explained by the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)

## Internal Issue (If applicable):
https://dev.azure.com/nventive/Uno%20Gallery/_workitems/edit/192996

### CODE REVIEW

File: src/library/Uno.Material/Controls/Chips/Chip.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+namespace Uno.Material.Controls
+{
+	public partial class Chip : ToggleButton
+	{
+		/// <summary>
+		/// Fires when a ToggleButton is checked or unchecked, except when set with <see cref="SetIsCheckedSilently"/>.
+		/// </summary>
+		internal event RoutedEventHandler IsCheckedChanged;
+
+		#region DependencyProperty: IsCheckable = true
```
Comment (MatFillion) [inline src/library/Uno.Material/Controls/Chips/Chip.cs:16]: I understand the name but I think we can find better =D

Comment (MatFillion) [reply src/library/Uno.Material/Controls/Chips/Chip.cs:16]: IsCheckEnabled ?

File: src/library/Uno.Material/Controls/Chips/Chip.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+			set => SetValue(IsCheckableProperty, value);
+		}
+
+		#endregion
+
+		private bool _isMuted = false;
+
+		public Chip()
+		{
+			Checked += RaiseIsCheckedChanged;
```
Comment (MatFillion) [inline src/library/Uno.Material/Controls/Chips/Chip.cs:39]: should we hook/unhook in `OnIsCheckableChanged`

File: src/library/Uno.Material/Controls/Chips/ChipGroup.cs [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+		}
+
+		#endregion
+
+		private bool _isSynchronizingSelection = false;
+		private bool _isUpdatingSelection = false;
+
+		public ChipGroup()
+		{
+#if WINDOWS_UWP
```
Comment (MatFillion) [inline src/library/Uno.Material/Controls/Chips/ChipGroup.cs:90]: it's not needed in Uno platforms? why? (possibly add comment to explain)

File: src/library/Uno.Material/Controls/Chips/ChipGroup.cs [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+		}
+
+		private void OnSelectionMemberPathChanged(DependencyPropertyChangedEventArgs e)
+		{
+			var binding = SelectionMemberPath != null
+				? new Binding { Path = new PropertyPath(SelectionMemberPath), Mode = BindingMode.TwoWay }
+				: null;
+			foreach (var container in GetItemContainers())
+			{
+				if (binding != null) container.SetBinding(Chip.IsCheckedProperty, binding);
```
Comment (MatFillion) [inline src/library/Uno.Material/Controls/Chips/ChipGroup.cs:114]: I think this line is worth of curly brackets ^^

File: src/library/Uno.Material/Controls/Chips/Chip.cs (hunk 2 above)
Comment (Xiaoy312) [reply src/library/Uno.Material/Controls/Chips/Chip.cs:39]: I'm avoiding [un]registering events unnecessarily. I believe the if-check approach would be less error-prone.

File: src/library/Uno.Material/Controls/Chips/Chip.cs (hunk 1 above)
Comment (Xiaoy312) [reply src/library/Uno.Material/Controls/Chips/Chip.cs:16]: That sounds wordy. Microsoft seems to agree with me:
https://www.google.com/search?q=ischeckable+site%3Adocs.microsoft.com
https://www.google.com/search?q=ischeckenabled+site%3Adocs.microsoft.com

File: src/library/Uno.Material/Controls/Chips/ChipGroup.cs (hunk 3 above)
Comment (Xiaoy312) [reply src/library/Uno.Material/Controls/Chips/ChipGroup.cs:90]: Uno exposes OnItemsSourceChanged, but the equivalent doesnt exist on uwp.
Actually, it should be cleaner to use just `RegisterPropertyChangedCallback` and forget `ItemsControl::OnItemsSourceChanged`.

File: src/library/Uno.Material/Controls/Chips/Chip.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls.Primitives;
```
Comment (MatFillion) [inline src/library/Uno.Material/Controls/Chips/Chip.cs:5]: Should we have a ICommand property on Chip? For the ActionChip

File: src/library/Uno.Material/Controls/Chips/Chip.cs (hunk 1 above)
Comment (MatFillion) [reply src/library/Uno.Material/Controls/Chips/Chip.cs:16]: Sold :)

File: src/library/Uno.Material/Controls/Chips/ChipGroup.cs (hunk 4 above)
Comment (Xiaoy312) [reply src/library/Uno.Material/Controls/Chips/ChipGroup.cs:114]: I think this should be fine for one-liner, but okay.

File: src/library/Uno.Material/Controls/Chips/Chip.cs (hunk 5 above)
Comment (Xiaoy312) [reply src/library/Uno.Material/Controls/Chips/Chip.cs:5]: `Chip` already offers it (via its `ButtonBase` ancestry).

**Comment by Xiaoy312** [comment]
Included in #402

**Pull request closed without merging**